diff -urN Mail-SpamAssassin-2.20.orig/spamassassin.raw Mail-SpamAssassin-2.20/spamassassin.raw
--- Mail-SpamAssassin-2.20.orig/spamassassin.raw	Wed Apr 10 04:57:56 2002
+++ Mail-SpamAssassin-2.20/spamassassin.raw	Mon Apr 22 14:54:58 2002
@@ -2,51 +2,6 @@
 
 use Getopt::Std;
 
-use FindBin;
-
-BEGIN {
-  # This code will track down the directories where WebMake
-  # keeps its modules, portably, so it'll work on Macs, UNIX and Win32,
-  # with or without a UNIX-style "make install" installation.
-  # Sadly, we can't rely on File::Spec to do the slash-twiddling for us;
-  # it's not included with some versions of MacPerl. :(
-  #
-  my $bin = $FindBin::Bin;
-  my $slash = '/';              # between directories in a path
-  my $dirtrailer = '';          # at the end of a directory's path
-
-  if ($^O eq 'MacOS') {
-    $slash = ':'; $dirtrailer = ':';
-  } elsif ($^O =~ /(win|os2)/) {
-    $slash = '\\';
-  }
-
-  # first, find the common candidates: "lib" and "site_perl" in
-  # the same dir as the script. These are likely on all platforms.
-  $_ = $bin.$slash. "lib" . $dirtrailer;
-  push (@INC, $_);
-  $_ = $bin.$slash. "site_perl" . $dirtrailer;
-  push (@INC, $_);
-
-  # next, support UNIX-style /usr-based installation, where the
-  # script lives in /usr/*/bin and the support files in /usr/*/lib
-  # or /usr/*/share. This only happens on UNIX afaik.
-  if ($slash eq '/') {
-    $_ = $bin . "/../lib/spamassassin";
-    if (-d $_) {
-      push (@INC, "$_/lib"); push (@INC, "$_/site_perl");
-    }
-
-    $_ = $bin . "/../share/spamassassin";
-    if (-d $_) {
-      push (@INC, "$_/lib"); push (@INC, "$_/site_perl");
-    }
-  }
-}
-
-use lib 'lib';
-use lib '../lib';		# for testing in ./t
-
 sub usage {
   my $ver = Mail::SpamAssassin::Version();
   my $rules = join ("\n\t\t\t\t", @Mail::SpamAssassin::default_rules_path);
